### PR TITLE
feat: two-tower neural CF on PyTorch

### LINKED ISSUE #23: feat: neural collaborative filtering (two-tower model)
A modern deep recommender behind an optional `torch` extra, on the Recommender interface.

**Acceptance criteria**
- Two-tower / matrix-factorization model that trains on MovieLens.
- Evaluated against the classical baselines in the benchmark.
- Tested on a tiny synthetic fixture so CI stays fast.

Blocked: needs the Recommender interface.

### PR DESCRIPTION
**Pull request by Burton-David**
Closes #23. The deep-learning headline algorithm: `recommender_systems.neural.TwoTowerCF` learns user and item embeddings via batched BPR-style ranking loss on PyTorch.

## Design

Same conceptual objective as the numpy `BPR` (#16) but lifted onto autograd, so the towers can grow side-information inputs or extra layers later without rewriting the training loop. Each user and item is an `nn.Embedding`; the score for a (user, item) pair is the dot product of their vectors; training maximizes `logsigmoid(pos_score − neg_score)` on observed-vs-random-negative pairs.

`TwoTowerCF` subclasses `_MatrixBackedRecommender` (#35), so the unknown-user fast path, seen-item mask, and top-n filter come free — `fit()` only has to populate `self._matrix` and `self._predicted`.

## Dependency story

`torch` is gated behind a new `neural` optional-dependency group. Without `[neural]`, `import recommender_systems.neural` raises a clear `ImportError` pointing at the extra:

```
pip install 'recommender-systems[neural]'
```

CI installs the extra on the 3.12 matrix leg only — `test_neural.py` opens with `pytest.importorskip("torch")` so the 3.10 and 3.11 legs skip the file. That keeps the matrix runtime sane while still verifying the neural path on every PR.

## Tests (6 new, 102 total when torch is present)

- Interface conformance.
- **Cluster recovery** on a synthetic 50-user two-cluster fixture — after 40 epochs of training, `recommend(user_0, n=10)` lands ≥8/10 items in the user's correct cluster. Same shape of test as `BPR`.
- Exclude-seen is enforced via the inherited `_MatrixBackedRecommender` recommend.
- Unknown-user fast path returns `[]`.
- Seeded fits are reproducible.
- Termination guard against a user who has rated every item (no infinite resample loop) — the same defensive pattern JJ flagged on `BPR` (#58).

## README

Algorithms table picks up the `TwoTowerCF` row; `BPR`'s note tightens to mention it's pure-numpy SGD, which makes the two distinct from a portfolio-reading perspective.

## Local checks

```
ruff check src tests scripts           # clean
ruff format --check src tests scripts  # clean
mypy                                   # Success: no issues found in 15 source files
pytest                                 # 102 passed
```

Closes #23

### CODE REVIEW

**Review by JohnJacob-coder — APPROVED** [review]
Thoroughly verified — this is the right way to add a heavy optional dep:
- torch is an optional [neural] extra (not core), the module guards its import with a helpful message, and it's NOT in __init__ — so 'import recommender_systems' never requires torch (confirmed: locally without torch the suite is 96 passed, 1 skipped).
- CI installs [neural] only on the 3.12 leg; test_neural.py uses pytest.importorskip so the other legs skip cleanly.
- Model is a correct two-tower with a BPR logsigmoid(pos-neg) loss; the has_negatives guard + vectorized resample can't hang, and you proactively added the all-items-user regression test (the lesson from #58 — nice).
- Tests verify it actually learns (cluster separation), reproducibility, seen-exclusion, and unknown users.
mypy/CI green. Excellent work.
